SWIPs/swip-draft_app_data_protocols.md

```diff
@@ -0,0 +1,119 @@
+
+Swarm can provide an opportunity to shift this relationship between data ownership and services using this data by allowing users to store their data directly in Swarm, and grant access to the apps and services they choose.
+The problem is then to define data formats that can be shared by different apps and services having different purposes, and protocols to ensure compatiblity and security between all the interested parties interacting with this data.
+
+Beyond end-user incentives of data ownership, shared formats and protocols can be beneficial to application and services developers as a way to get started faster with development on top of Swarm, and possibly access to existing user data.
+
+## Specification
+
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for the current Swarm platform and future client implementations.-->
+
+### Scope (WIP)
+
+- Should only rely on existing features of Swarm, but might evolve as new possibilities are added
+- Should be usable via a HTTP gateway
+- Should protect the user's security and privacy as much as possible, and document necessary trade-offs
+- Should define core data structures and validation methods
+- Should define custom extensions to core data structures
+- Should define protocols for data authoring and discovery accross multiple apps/devices/services and between different users
+- Should support versioning
+
+### Terminology (WIP)
+
+- Actor: human entity (person, group...)
+- Agent: code acting on behalf of an Actor (app, device, service...)
+- Resource: data accessible on Swarm
+  - File (any binary data)
+  - Entity (JSON data defined in spec)
+  - Source (single Entity or Entity feed)
+  - Publication (list of Sources)
+
+### Research areas (WIP)
+
+#### Data structure format
+
+- Should we use JSON or a binary format such as protocol buffers?
+- What about validation?
+
+#### Entities extensibility
+
+- Can we restrict the spec to limited number of core Entities while allowing for more complex Entities to be added as extensions?
+- Can we provide fallback types for complex Entities that would not be supported by a given client?
+- What about validation of unknown (non-core or from supported extensions) Entities?
+
+#### Core entities and protocols
+
+- Define what entities should be part of core vs extensions.
+- Define validation rules for these core entities.
+- Define discover and exchange rules for these entities.
+
+#### Key management
+
+- How could an Actor add and remove (revoke) Agents and Resources at will?
+- How could an Agent add and remove (if own) a Source from a Publication?
+
+#### Data encryption
+
+- Should a single algorithm or multiple ones be supported/recommended?
+- Should Swarm built-in encryption and/or ACT be used?
```

Comment:
I would prefer to use light nodes everywhere and use Swarm's built-in encryption. However light nodes are not going to be production ready for a while and there may be always use-cases where they are not feasible to use (think of embedded devices). Also I can imagine hybrid products and solutions that are not completely decentralized but are using Swarm as a storage layer and therefore using gateways to access data.
For those cases it would be good to have a protocol level, standardised encryption layer to provide end-to-end encryption. I was thinking of somehow using Swarm's encryption code for that but the encryption in Swarm is tightly coupled with how the files are stored so I assume it would come with a big dependency on other components (think of chunker, hashing etc).

Reply:
I agree light nodes are the best option but I think it would be good to start working on these protocols based on the current version of Swarm so we are not limited by Swarm's development timeline, though obviously as new features are released we should adapt accordingly.

I think the question is also if we want to support Web apps hosted on Swarm and accessed using a Web browser via a Swarm gateway?
Personally I'd rather not support this use case because I think there are too many potential security and privacy risks, but at the same time if this is a primary use-case for Swarm I think these protocols should support it in order to be relevant.

Reply:
Well, again it's the question of adoption. Assuming installing something decreases the willingness of people to try out new things, I think most people will use their first dapps from the browser. And until the majority of the browsers don't support Swarm the only option is a gateway.

I also understand the risks you mentioned with this but if we already have implementations in Javascript I don't see a reason why the browsers should not be supported. Regarding the protocol this doesn't require too much effort on our side, for example in the case of encryption the only requirement is to choose a standard which has a Javascript implementation which is also supported in the browser.

What do you think?

Reply:
Yes I think it's OK to be cautious about not defining requirements that would prevent usage in Web browsers, I guess I'm mostly concerned about developers hosting their apps using a gateway, that would completely undermine security and privacy measures the protocols would otherwise define.